### PR TITLE
Add class type for RzTypes

### PR DESCRIPTION
**Pull request by Basstorm**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

When integrating `PDB` information into `RzTypes`, There is no class type for now to store.

**Test plan**

...

**Closing issues**

Partially addresses https://github.com/rizinorg/rizin/issues/1205

### CODE REVIEW

File: librz/include/rz_type.h [hunk 1]
```diff
@@ -70,6 +70,7 @@ typedef enum {
 
 typedef enum {
 	RZ_BASE_TYPE_KIND_STRUCT,
+	RZ_BASE_TYPE_KIND_CLASS, // Just a "placeholder" for now.
```
Comment (XVilka) [inline librz/include/rz_type.h]: I think it's better to put it after ATOMIC one.

File: librz/include/rz_type.h [hunk 2]
```diff
@@ -88,6 +89,13 @@ typedef struct rz_type_struct_member_t {
 	size_t size; // in bits?
 } RzTypeStructMember;
 
+typedef struct rz_type_class_member_t {
+	char *name;
+	RzType *type;
+	size_t offset; // in bytes
+	size_t size; // in bits?
```
Comment (XVilka) [inline librz/include/rz_type.h]: Remove `?` here.

File: librz/include/rz_type.h [hunk 3]
```diff
@@ -88,6 +89,13 @@ typedef struct rz_type_struct_member_t {
 	size_t size; // in bits?
```
Comment (XVilka) [inline librz/include/rz_type.h]: And here while you are on it

File: librz/include/rz_type.h [hunk 4]
```diff
@@ -115,6 +127,7 @@ typedef struct rz_base_type_t {
 	RzTypeAttribute attrs;
 	union {
 		RzBaseTypeStruct struct_data;
+		RzBaseTypeClass class_data;
 		RzBaseTypeEnum enum_data;
 		RzBaseTypeUnion union_data;
```
Comment (XVilka) [inline librz/include/rz_type.h:131]: Better to put it after `union_data`

File: librz/include/rz_type.h [hunk 5]
```diff
@@ -133,6 +146,7 @@ typedef enum {
 typedef enum {
 	RZ_TYPE_IDENTIFIER_KIND_UNSPECIFIED,
 	RZ_TYPE_IDENTIFIER_KIND_STRUCT,
+	RZ_TYPE_IDENTIFIER_KIND_CLASS,
 	RZ_TYPE_IDENTIFIER_KIND_UNION,
 	RZ_TYPE_IDENTIFIER_KIND_ENUM
```
Comment (XVilka) [inline librz/include/rz_type.h]: Same here, after `KIND_ENUM`

File: librz/type/base.c [hunk 6]
```diff
@@ -18,6 +18,13 @@ RZ_API void rz_type_base_struct_member_free(void *e, void *user) {
 	free((char *)member->name);
 }
 
+RZ_API void rz_type_base_class_member_free(void *e, void *user) {
```
Comment (XVilka) [inline librz/type/base.c]: Move this function after `union_member_free()`

File: librz/type/base.c [hunk 7]
```diff
@@ -117,6 +124,9 @@ RZ_API void rz_type_base_type_free(RzBaseType *type) {
 	case RZ_BASE_TYPE_KIND_STRUCT:
 		rz_vector_fini(&type->struct_data.members);
 		break;
+	case RZ_BASE_TYPE_KIND_CLASS:
```
Comment (XVilka) [inline librz/type/base.c]: Same, at the end, after all C types

File: librz/type/base.c [hunk 8]
```diff
@@ -147,6 +157,9 @@ RZ_API RZ_OWN RzBaseType *rz_type_base_type_new(RzBaseTypeKind kind) {
 	case RZ_BASE_TYPE_KIND_STRUCT:
 		rz_vector_init(&type->struct_data.members, sizeof(RzTypeStructMember), rz_type_base_struct_member_free, NULL);
 		break;
+	case RZ_BASE_TYPE_KIND_CLASS:
```
Comment (XVilka) [inline librz/type/base.c]: And here

File: librz/type/base.c [hunk 9]
```diff
@@ -193,6 +206,17 @@ RZ_API RZ_OWN char *rz_type_db_base_type_as_string(const RzTypeDB *typedb, RZ_NO
 		rz_strbuf_append(buf, " }");
 		break;
 	}
+	case RZ_BASE_TYPE_KIND_CLASS: {
```
Comment (XVilka) [inline librz/type/base.c]: And here

**Review by XVilka — CHANGES_REQUESTED** [review]

**Comment by XVilka** [comment]
Can you add a test in `test/unit/test_types.c` also? We cannot parse them yet, but surely will be able to form them manually.
Also we need to think about serialization/deserialization. cc @thestr4ng3r

File: librz/include/rz_type.h [hunk 10]
```diff
@@ -74,6 +74,7 @@ typedef enum {
 	RZ_BASE_TYPE_KIND_ENUM,
 	RZ_BASE_TYPE_KIND_TYPEDEF,
 	RZ_BASE_TYPE_KIND_ATOMIC, // For real atomic base types
+	RZ_BASE_TYPE_KIND_CLASS, // Just a "placeholder" for now.
```
Comment (XVilka) [inline librz/include/rz_type.h]: I don't think we need to mark it as a placeholder. Some basic features will work already.

**Review by XVilka — APPROVED** [review]
Looks good now. But please add a test.

**Review by wargio — APPROVED** [review]

File: test/unit/test_type.c [hunk 11]
```diff
@@ -488,6 +488,146 @@ static bool test_struct_func_types(void) {
 	mu_end;
 }
 
+static bool test_form_class_type_maually(void) {
```
Comment (XVilka) [inline test/unit/test_type.c]: `s/maually/manually/g`

File: librz/include/rz_type.h [hunk 12]
```diff
@@ -85,20 +86,31 @@ typedef struct rz_type_struct_member_t {
 	char *name;
 	RzType *type;
 	size_t offset; // in bytes
-	size_t size; // in bits?
```
Comment (XVilka) [inline librz/include/rz_type.h:88]: I was actually wondering, if we switch `offset` measure unit from `bytes` to `bits` we will get the [bitfields support](https://github.com/rizinorg/rizin/issues/1240) for free. @Basstorm @thestr4ng3r @ret2libc @wargio what do you think?

**Review by XVilka — APPROVED** [review]

**Review by thestr4ng3r — COMMENTED** [review]
I am not sure if this overall is a good idea. Classes and types are conceptually very different. Normally you will have a class and optionally an associated type (which then would be just a struct). For example you can also have a class that has no type at all, like a pure abstract one or one with only static members.
That is the idea behind analysis classes, which already exist and could be extended to refer to types too.

File: librz/include/rz_type.h (hunk 12 above)
Comment (wargio) [reply librz/include/rz_type.h:88]: yes, but change the definition to ut64, so we are sure that is really an ut64

Comment (thestr4ng3r) [reply librz/include/rz_type.h:88]: sgtm

**Comment by ret2libc** [comment]
> I am not sure if this overall is a good idea. Classes and types are conceptually very different. Normally you will have a class and optionally an associated type (which then would be just a struct). For example you can also have a class that has no type at all, like a pure abstract one or one with only static members.
That is the idea behind analysis classes, which already exist and could be extended to refer to types too.

@XVilka was this addressed anywhere? What is the decision here? Is this PR still necessary?

**Comment by XVilka** [comment]
@ret2libc we will prepare the different PR with a different approach later, but please keep this open until we open the new one. There is some useful stuff in it.

**Comment by XVilka** [comment]
We decided to make this a bit differently.

**Pull request closed without merging**